addons/website_forum/tests/test_controller.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+from odoo.tests.common import HttpCase, new_test_user
+
+
+class TestController(HttpCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        portal_user = new_test_user(cls.env, login='portal_user', groups='base.group_portal')
+        cls.portal = portal_user.login
+        admin_user = new_test_user(cls.env, login='admin_user', groups='base.group_user,base.group_system')
+        cls.admin = admin_user.login
+        cls.headers = {"Content-Type": "application/json"}
+        cls.pixel = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAADElEQVR4nGNgYGAAAAAEAAH2FzhVAAAAAElFTkSuQmCC'
+
+    def _build_payload(self, params=None):
+        """
+        Helper to properly build jsonrpc payload
+        """
+        return {
+            "jsonrpc": "2.0",
+            "method": "call",
+            "id": 0,
+            "params": params or {},
+        }
+
+    def test_portal_attachment_01(self):
+        self.authenticate(self.portal, self.portal)
+        payload = self._build_payload({"name": "pixel", "data": self.pixel, "is_image": True, "res_model": "forum.post"})
+        response = self.url_open('/web_editor/attachment/add_data', data=json.dumps(payload), headers=self.headers)
+        self.assertEqual(200, response.status_code)
+        attachment = self.env['ir.attachment'].search([('name', '=', 'pixel')])
+        self.assertTrue(attachment)
```

Comment:
Why not also checking the fetch behavior for portal? (You check if for admin but not for portal, either check both or just portal)

Note that you can probably just encapsulate the code to avoid a copy paste (always bad to maintain).

```py
def test_01_xxx(self):
    def _xxx(attach_name):
        build_payload
        url_open(create(attach_name))
        assert(200)
        search([])
        assertTrue
        build_payload
        url_open(fetch(attach_name))
        assert(200)
        assertin(attach.id)

    self.authenticate(self.portal, self.portal)
    _xxx('test portal attach')
    self.authenticate(self.admin, self.admin)
    _xxx('test admin attach')
```

Reply:
The fetch behavior for portal users returns an empty list and attachments created on forum posts are private (see https://github.com/odoo/odoo/pull/82612#discussion_r851153823), so we can't test the fetch for portals and for admin, we can't fetch the attachment that was uploaded using the controller.